New Processing Algorithm: Equidistance Buffer

spatialthoughts opened the pull request:
This is a new algorithm that approximate geodesic buffers in QGIS by transforming each feature geometry to a custom Azimuthal Equidistance projection centered at that geometry, and then transforming it back to the original projection. I have written a post explaining the rationale behind this algorithm https://spatialthoughts.com/2019/04/05/geodesic-buffers-in-qgis/

elpaso committed:
Handle the coordinate tranform context in the layer ctor

Also use the context when read/write XML by passing it
in the R/W context

elpaso committed:
coordinateTransformContext -> transformContext for consistency

I like "coordinateTransformContext" more btw.

elpaso committed:
More coordinateTranformContext -> tranformContext

elpaso committed:
Remove deprecated QgsCoordinateTransform ctor ...

... and add a getter for coordinateTransform to QgsMapLayer

The reason is that I think we should reduce the dataProvider API
usage and rely on the higher level QgsMapLayer API whenever it
is possible, QgsMapLayer checks for data provider validity
and returns a default constructed QgsCoordinateTransform instead
of crashing.

elpaso committed:
Dox typos

elpaso committed:
Removed leftover

elpaso committed:
Use options from provider ctor

elpaso committed:
Deprecate vector layer ctor with default constructed options

replaced by QgsVectorLayer( const QgsVectorLayer::LayerOptions &options,
...

elpaso committed:
Add mandatory options to QgsVectorLayer ctor

... and deprecate the rest.

In QGIS 4 we'll have to get rid of the deprecated default ctor and
create a new one which takes no arguments.

elpaso committed:
Deprecate method without transformContext

elpaso committed:
Deprecate default ctor for mesh layer options

nyalldawson committed:
Tweak QgsProcessingUtils::combineLayerExtents for future proofing, remove deprecated usage

nyalldawson committed:
Fix use of deprecated/removed coordinate transform constructor

nyalldawson committed:
Some build fixes

jef-n committed:
fix #21872 (refs #21774; partly reverts e57d1e06f0 and 507e327c0c)

src/python/qgspythonutilsimpl.cpp:596 produces "lost sys.stdout" when
qgis.utils.updateAvailablePlugins() returns.

elpaso committed:
Fix test build

elpaso committed:
Restore default ctor for options and layer

elpaso committed:
doxy

elpaso committed:
Do not store context in layer (but in provider) ...

... also use transform context in raster projector
and deprecate setCrs without context

elpaso committed:
Raster transform context

elpaso committed:
Fix class documentation

elpaso committed:
Missing doc in an unrelated class :(

elpaso committed:
More doxy

elpaso committed:
Remove build warnings

ujavalgandhi committed:
New algorithm EquidistanceBuffer

See detailed explanation at https://spatialthoughts.com/2019/04/05/geodesic-buffers-in-qgis/

ujavalgandhi committed:
New algorithm EquidistanceBuffer

See detailed explanation at https://spatialthoughts.com/2019/04/05/geodesic-buffers-in-qgis/

ujavalgandhi committed:
New algorithm EquidistanceBuffer

See detailed explanation at https://spatialthoughts.com/2019/04/05/geodesic-buffers-in-qgis/

ujavalgandhi committed:
New algorithm EquidistanceBuffer

See detailed explanation at https://spatialthoughts.com/2019/04/05/geodesic-buffers-in-qgis/

ujavalgandhi committed:
Rename EquidistanceBuffer to EquidistanceBuffer.py

ujavalgandhi committed:
Rename EquidistanceBuffer to EquidistanceBuffer.py

dmarteau committed:
Set config path before calling plugin method

elpaso commented on python/plugins/processing/algs/qgis/EquidistanceBuffer.py:
```diff
@@ -0,0 +1,142 @@
+from processing.algs.qgis.QgisAlgorithm import QgisFeatureBasedAlgorithm
+
+
+class EquidistanceBuffer(QgisFeatureBasedAlgorithm):
```
Cool, thanks! 
May I suggest that you copy the alg description in the class docstring?

spatialthoughts replied:
Thanks. I have added the docstring

ujavalgandhi committed:
Add class docstring

ujavalgandhi committed:
Add class docstring

ujavalgandhi committed:
Add class docstring

luipir commented on python/plugins/processing/algs/qgis/EquidistanceBuffer.py:
```diff
@@ -0,0 +1,142 @@
+                            'Miter',
+                            'Bevel']
+        self.end_cap_styles = [self.tr('Round'),
+                       'Flat',
```
why translate only 'Round' ?

spatialthoughts replied:
I started from VariableDistanceBuffer.py which had only Round translated. Fixed now

luipir commented on python/plugins/processing/algs/qgis/EquidistanceBuffer.py:
```diff
@@ -0,0 +1,142 @@
+
+    def initParameters(self, config=None):
+        self.addParameter(QgsProcessingParameterDistance(self.DISTANCE,
+                                                         self.tr('Distance (in meters'),
```
syntax error closing the ) in the wrong place?

spatialthoughts replied:
I don't see it. The script runs without error too. Can you specify where is the syntax error?

luipir replied:
```suggestion
                                                         self.tr('Distance (in meters)'),
```
you ar right is not a syntax error is a simple typo

ujavalgandhi committed:
Translate all join and end cap styles

ujavalgandhi committed:
Translate all join and end cap styles

ujavalgandhi committed:
Translate all join and end cap styles

luipir commented:
tnx @spatialthoughts for the core contribution :) may you add also a processing test for this alg? it's is especially important for core algs.

luipir committed:
Update python/plugins/processing/algs/qgis/EquidistanceBuffer.py

Co-Authored-By: spatialthoughts <ujaval@gmail.com>

luipir committed:
Update python/plugins/processing/algs/qgis/EquidistanceBuffer.py

Co-Authored-By: spatialthoughts <ujaval@gmail.com>

luipir committed:
Update python/plugins/processing/algs/qgis/EquidistanceBuffer.py

Co-Authored-By: spatialthoughts <ujaval@gmail.com>

ujavalgandhi committed:
Add a test for equidistance buffer

ujavalgandhi committed:
Added expected test output files

spatialthoughts commented:
> tnx @spatialthoughts for the core contribution :) may you add also a processing test for this alg? it's is especially important for core algs.

Thanks for the quick review @luipir . I have added the test. This is the first time doing a processing test, so let me know if this looks ok.

nyalldawson commented on python/plugins/processing/algs/qgis/EquidistanceBuffer.py:
```diff
@@ -0,0 +1,148 @@
+                       self.tr('Square')]
+
+    def initParameters(self, config=None):
+        self.addParameter(QgsProcessingParameterDistance(self.DISTANCE,
```
You can call setDefaultUnit on this parameter with a unit type of meters to improve the ux

spatialthoughts replied:
Nice! I was annoyed by the <unknown> label. This is much better. I have added it.

luipir commented:
@spatialthoughts other than indentation seems you missed to add the alg in some cmake... AFAIK this would be the travis fail

luipir commented:
and... tnx for the tests

ujavalgandhi committed:
Set default unit on distance parameter

ujavalgandhi committed:
Set default unit on distance parameter

ujavalgandhi committed:
Set default unit on distance parameter

ujavalgandhi committed:
fix pep8 errors

ujavalgandhi committed:
fix pep8 errors

ujavalgandhi committed:
fix pep8 errors

spatialthoughts commented:
@luipir I have fixed the pep errors and now there is only 1 failing check. But I don't understand which cmake I missed. Any insights would be helpful.

Gustry commented:
@luipir It's python, there is no cmake.

@spatialthoughts Your error is here https://travis-ci.org/qgis/QGIS/jobs/518718078#L2936

luipir commented:
seems the registry can't find your alg... can't find why in this moment, it seems all correctly set

m-kuhn commented:
add the alg to this list too https://github.com/qgis/QGIS/blob/8fa868a2e72a2af6b1b1ce70f1ac2cff65bae323/python/plugins/processing/algs/qgis/QgisAlgorithmProvider.py#L160

ujavalgandhi committed:
Add layer name

spatialthoughts commented:
Thanks @m-kuhn . The test runs now but fails with a strange error. I don't know why it is trying to create a gpkg here

Processing(2): Could not create layer /tmp/tmploce9n5a/equidistance_buffer_points.gml'.gpkg: Creation of field fid failed (OGR error: Wrong field type for fid)
Processing(2): Could not create layer /tmp/tmploce9n5a/equidistance_buffer_points.gml'.gpkg: Creation of field fid failed (OGR error: Wrong field type for fid)

m-kuhn commented on python/plugins/processing/algs/qgis/EquidistanceBuffer.py:
```diff
@@ -0,0 +1,168 @@
+        self.source_crs = source.sourceCrs()
+        if not self.source_crs.isGeographic():
+            feedback.reportError(
+                'Layer CRS must be a Geograhpic CRS for this algorithm')
```
Should this be translated?

spatialthoughts replied:
Yes. Done.

ujavalgandhi committed:
Update EquidistanceBuffer.py

ujavalgandhi committed:
Update EquidistanceBuffer.py

ujavalgandhi committed:
Update EquidistanceBuffer.py

DelazJ commented on python/plugins/processing/algs/qgis/EquidistanceBuffer.py:
```diff
@@ -0,0 +1,169 @@
+    """
+    This algorithm creates equidistance buffers for vector layers.
+
+    Each geometry is transformed to a Azimuthal Equidistant projection centered
```
Is it **a** Azimuthal or **an** Azimuthal?

spatialthoughts replied:
I wrote it thinking 'a' refers to the projection and not azimuthal. A "Azimuthal equidistant' projection. I am not sure which is right though.

DelazJ replied:
> I am not sure which is right though.

me neither ;)

roya0045 replied:
An, if a vowel is after 'a ' it becomes 'an '.

spatialthoughts replied:
Thanks. @roya0045 and @DelazJ . I have replaced 'a' with 'an'.

DelazJ commented on python/plugins/processing/algs/qgis/EquidistanceBuffer.py:
```diff
@@ -0,0 +1,169 @@
+        if not self.source_crs.isGeographic():
+            feedback.reportError(
+                self.tr(
+                    'Layer CRS must be a Geograhpic CRS for this algorithm'))
```
```suggestion
                    'Layer CRS must be a Geographic CRS for this algorithm'))
```
and do we need the `for this algorithm` part?

m-kuhn replied:
good point, it should be clear from the context that it concerns "this algorithm"

spatialthoughts replied:
fixed

ujavalgandhi committed:
Fix typo

ujavalgandhi committed:
Fix typo

ujavalgandhi committed:
Fix typo

rldhont committed:
Test apply SLD PointSymbolizer on single centroid for polygon

elpaso committed:
Fix grass alg combineLayerExtents

spatialthoughts commented:
I am still not sure what is causing the travis build to fail. The log has some entries which indicate that the test is trying to create geopackage and failing, but I don't know why the gml file would be converted to gpkg. Any suggestions?

elpaso committed:
Remove test case (API is not public anymore)

nyalldawson committed:
Update src/core/layout/qgslayoutitemmapoverview.cpp

Co-Authored-By: elpaso <elpaso@itopen.it>

nyalldawson committed:
Update src/core/mesh/qgsmeshlayer.h

Co-Authored-By: elpaso <elpaso@itopen.it>

nyalldawson committed:
Update src/core/qgsmaplayer.h

Co-Authored-By: elpaso <elpaso@itopen.it>

elpaso committed:
Pass transform context to raster calculator

elpaso committed:
Use transform context from rw context

elpaso committed:
Drop comment

elpaso committed:
Use new QgsRasterCalculator ctor with transform context

elpaso committed:
doxy

nirvn committed:
Add QgsGuiUtils::iconSize to get user-prefered toolbar icon size

nirvn committed:
[ui] Fix data source selection dialog icon size

nyalldawson committed:
Fix incorrect annotation scaling when exporting layouts

Previously, annotation size and position always used pixel units. This
did not work well when exporting layouts, resulting in tiny annotations
(it also caused issues when moving projects between hidpi/non hidpi
displays).

Instead, use millimeters for annotation size and position so that the
appearance is consistent across displays and works correctly in layout
exports.

Add lots of unit tests covering this.

Fixes #18373

elpaso committed:
Pass transformContext to raster writer

Commit:
[QgsQuick] Updated external resource widget
It has been enlarged, button changed
Added a new component qgsquickicontextitem

elpaso committed:
Update layer transform context when adding them to a project

elpaso committed:
Set transform context from layer options

elpaso committed:
Get transform context from project instance in app

elpaso committed:
Add transform context to raster file writer task

elpaso committed:
Unit tests for transform context project<->layer sync

elpaso committed:
Fix doxy deprecated

elpaso committed:
Don't use deprecated fileWriter.writeRaster

Commit:
[QgsQuick] Customizable icons for external source widget
..and date time widget. Changes in other widgets due to passing a whole style object to components.

elpaso committed:
Get transformContext from layer instead of provider

Commit:
[QgsQuick] Styling of tab buttons in feature form
Added properties for color background setting

nirvn committed:
[ui] Fix layout designer panel icon (re)sizing

nirvn committed:
[themes] Avoid hardcoded black for default console timestamp colors
(fixes #21860)

nirvn committed:
[themes] Improve night mapping hyperlink color

nirvn committed:
[processing] Add 'distance' tag to join attributes by nearest alg.

PeterPetrik committed:
fix rendering of mesh triangle with 0,0 vertex

PeterPetrik committed:
update MDAL 0.3.1

Commit:
[QgsQuick] Added styling properties to feature form tabs

elpaso committed:
[server] Implement LAYERFONTCOLOR and ITEMFONTCOLOR

Fixes #21871
Fixes #21870

Funded by: Gis3w

elpaso committed:
Layout

elpaso committed:
Waste my time with Travis

alexbruy committed:
[processing] ensure that output temporary directory exists (fix #21701)

m-kuhn committed:
Add @snapping_results variable for points

Fix #21784 https://issues.qgis.org/issues/21784

alexbruy committed:
[processing] populate batch interface with rows when multiple layers
selected (fix #21859)

m-kuhn committed:
Add a comment to explain the reason for the code

roya0045 committed:
More verbose sip failure

jef-n committed:
fix 3d7c711db (refs #21872)

nyalldawson committed:
Add API for specifying fallback CRS and layer WKB type to use
for QgsVectorLayer, when constructing a layer with a URI which
may be invalid

These may be used for layers where the geometry type/CRS is
known in advance, and where the layer path may not be
initially resolvable. (E.g. layers with a URI pointing to a
non-existant file). It is only ever used if the layer cannot
be resolved, otherwise the actual layer geometry type will be
detected and used for the layer.

nyalldawson committed:
Add unit test for f947ffb5b

nyalldawson committed:
Store vector layer wkb type in xml

We can use this when restoring the layer, if the uri turns out
to be invalid at that stage (e.g. a file has moved). By storing
and falling back to the last known wkb type, we avoid unnecessarily
discarding the existing layer renderer, and can still show the
expected layer type in the layer tree.

elpaso committed:
Fix wrong cotrol image for ITEMFONTCOLOR test

nirvn committed:
[ogr] Only throw validity error in reloadData() when previously valid

elpaso committed:
Initial check pk unicity

elpaso committed:
Tooltips for PG options in source select dialog

Commit:
[QgsQuick] Handle empty config for checkbox editor widget
Use true/false instead of un/checkedState if empty

Commit:
[QgsQuick] Checkbox widget fix
Correct switching on/off

rldhont committed:
[Bugfix][Server] WMTS: read the service url from project

It is a typo fix

rldhont committed:
[Bugfix][Server] WMTS: CRS can have axis inverted

The top left element has to respect the axis, like in WMS 1.3.0.

rldhont committed:
[Bugfix][Server] WMTS - use resolution for bbox calculation

Because of the limit of double size in C++, it's bettre to use the resolution instead of the scale denominator for calculating tiles extent.

rldhont committed:
[Tests][Server] WMTS - use resolution for bbox calculation

spatialthoughts commented:
@luipir @nyalldawson I would love to merge this, but not able to figure out how to fix the test. Any pointers?

luipir commented on python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml:
```diff
@@ -1386,6 +1386,22 @@ tests:
+      SEGMENTS: 8
+    results:
+      OUTPUT:
+        name: expected/equidistance_buffer_points.gml'
```
as in https://travis-ci.org/qgis/QGIS/jobs/519176767#L1542 you have a ' more at the end of filename. Not clear to me is it's enough to fix the issue
```suggestion
        name: expected/equidistance_buffer_points.gml
```

luipir committed:
Update python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml

Co-Authored-By: spatialthoughts <ujaval@gmail.com>

luipir committed:
Update python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml

Co-Authored-By: spatialthoughts <ujaval@gmail.com>

wonder-sk committed:
[3d] Fix unnecessary terrain map updates when changing 3D renderer

When using layer styling dock, every change in 3D rendering configuration
was also triggering update of 2D map which also forces update of all
terrain tile textures with a new 2D map which wasn't really needed.
The fix makes the triggerRefresh() call on layer optional - each layer
styling dock config widget can tell whether its updates require 2D map
refresh (true by default).

luipir commented:
hmmmm... here seems a testing environment error
https://travis-ci.org/qgis/QGIS/jobs/522770593#L1541
I suppose executing: 
https://github.com/qgis/QGIS/pull/9774/files#diff-7add1e33bfdf87577a677df3907a62f4R157
and from there there is the result mismatch.

I'm not sure I supposed well, btw I never found this kind of error. @nyalldawson @m-kuhn did you?

nirvn committed:
Turn antialising render flag on when appropriate in QgsRenderContext::fromQPainter()

nirvn committed:
Turn antialiasing flag on when rendering legend symbol preview

wonder-sk committed:
[3d] fix shading of 3D point symbols

The vertex shader was incorrectly transforming world position and normal
vectors for phong shading. This was causing wrong appearance that was noticeable
when moving/rotating camera (e.g. specular reflection was always in the same place).
This was caused by involving view matrix (i.e. camera position and orientation)
to the vectors, but the world position and normal should be independent from camera.

wonder-sk committed:
[3d] Fix edge drawing when multiple renderer rules have edges enabled

When using rule-based 3D rendering, if multiple rules were enabled with edge rendering,
only the first rule was getting correctly set up line width for line material.
(the rest were getting some very thick line width and making the 3D view unusable)

Commit:
[QgsQuick] Added callbacks for qgsquick external source widget
Added callbacks functions in external resource handler - if some actions as permanent removal of image suppose to be done on form save.
Some changes in design

jef-n committed:
debian packaging: include python3-lxml

dmarteau committed:
Remove plugin module path  when unloading

    Unloading plugin must remove added module path
    in order to prevent wrong package import when reloading

nirvn committed:
[processing] Log features that could not be divided in tessellate alg.

nirvn committed:
[FEATURE][symbology] Add offset settings for point pattern fill symbol layer

Commit:
Initial version of xyz tiles export (processing tool)

Commit:
Registration of TilesXYZ algorithm

signedav committed:
fix comment since 3.4

Commit:
Commented out code for debugging (saving of metatiles), fixed python formatting

Commit:
[QgsQuick] External widget changes
Simplified callbacks for onSaved/onCanceled form
Some design fixes.

elpaso committed:
[postgres] Fix checkPrimaryKeyUnicity option

This provider option was linked to the project level option
"Trust layer metadata..." which was implemented
to speed up loading of large dataset by trusting extent
read from metadata to avoid costly operations to determine
the layer extent.

Check PK unicity on the other hand has only effect on views
and query layers and it is useful as an independent
option to prevent loading of layers that have no PK (or the
wrong one).

But the operation of determine unicity of a values in a column
can also be costly, so better to get control back to the user.

Legacy default is preserved (the project-level "Trust..." option).

Fixes #21839

Funded by RAAB.nl

Commit:
[QgsQuick] External resource widget changes
Enlarge buttons and modified text for buttons

nirvn committed:
Add forgotten since tags on new point pattern fill functions

ujavalgandhi committed:
Update EquidistanceBuffer.py

ujavalgandhi committed:
Update EquidistanceBuffer.py

ujavalgandhi committed:
Update EquidistanceBuffer.py

ujavalgandhi committed:
Update qgis.yaml

Commit:
MBTilesWriter can write tiles in any order, proper transforming of extents

Commit:
[QgsQuick] Handling missing config values for CheckBox widget
Added widget's properties for un/checked value - if a config is missing for boolean field, default value is set and work with. If those values are not set for a different type of a field, its considered as not valid setup and will not be working.

Commit:
Adjusted min/max/default values of parameters, moved some code into 'prepareAlgorithm'

Commit:
Added help text

alexbruy committed:
add missed place marker to the translatable string

wonder-sk committed:
[processing] Add 'project' to test definition and 'directory' output test

- for algorithms that produce directory output, it is possible to test
  that directory contents are exactly the same (recursively)
- added possibility to have a project file loaded before an algorithm is run
- documented the new additions (+ few existing ones)

wonder-sk committed:
Remove NAME parameter, use OUTPUT_FILE instead

For output to directory, OUTPUT_DIRECTORY destination parameter is used.
For output to MBTiles file, OUTPUT_FILE destimation parameter is used.

wonder-sk committed:
Add test for XYZ Tiles algorithm

wonder-sk committed:
Raise an exception if the necessary parameter was not specified

nyalldawson committed:
Add missing /Transfer/ annotation

nyalldawson committed:
Fix exception when reloading plugins

nyalldawson committed:
[processing] Fix exception on completion of batch execution of
algorithm which has string or numeric outputs

nyalldawson committed:
[processing] Correctly reset progress bar after completing batch execution

nyalldawson committed:
Remove some unused code

nyalldawson committed:
[processing][needs-docs] Move batch execution double-click-to-fill-down
to an explicit widget in the table

The double-click-header action is very hidden, and many users will
not stumble upon this. By moving it to an explicit "Autofill" widget at the
top of the table, we make this important action much more user
discoverable.

It also gives the possibility of other, parameter specific, autofill
actions (e.g. fill by expression, fill by file pattern)

nyalldawson committed:
Add widget and dialog for finding files by pattern match, with an
optional recursive search

nyalldawson committed:
[FEATURE][processing] In batch mode, allow population of file/layer
input columns by searching for files matching a specified pattern

With optional recursive search!

nyalldawson committed:
Indentation

elpaso committed:
More tests for checkPrimaryKeyUnicity

PeterPetrik committed:
fix #21931 Mesh calculator crash when used with face dataset

elpaso committed:
Layout and SQL

nirvn committed:
[symbology] Fix custom dash pattern for line with an hairline (i.e. 0)  width

elpaso committed:
[server] Respect custom layer order for groups in GetMap

Fixes #21917

elpaso committed:
Suggestions from JEF

elpaso committed:
Fix identify for WMS client with json format

Fixes #21735

elpaso committed:
Respect number formatting in identify raster results

troopa81 committed:
Fix point picking precision for point parameters in processing dialog

Commit:
[QgsQuick] Modified Checkbox widget
Replaced checkbox with switch.
Registered QVariant::Type to recognize field type. When value was empty/not set, it wasn't clear what is a type of field. It caused wrong behavior while setting default un/checked state.

nyalldawson committed:
Refactor creation of batch dialog row algorithm parameters

nyalldawson committed:
[FEATURE][processing] Add option to calculate parameter values by expression in batch dialog

This new option is available under the Autofill menu for a column. Selecting it allows users
to create a new QGIS expression to use to update the value inside that column. Existing
parameter values (including those from other columns) are available for use inside
the expression via @variables.

E.g. this allows setting output file names to complex expressions like
'/home/me/stuff/buffer_' || left(@input, 30) || '_' || @distance || '.shp'

nyalldawson committed:
[FEATURE][processing] New "Add values by expression" option for
populating batch processing dialog

This option adds news rows using the values from an expression
which returns an array. (As opposed to "Calculate by Expression",
which works only on existing rows).

The intended use case is to allow populating the batch dialog
using complex numeric series, e.g. those created by the "generate_series"
expression function.

For example, adding rows for a batch buffer using the expression

    generate_series(100, 1000, 50)

(results in new rows with values 100, 150, 200, .... 1000)

nyalldawson committed:
[FEATURE] Expression functions which return file info

base_file_name: Returns the base name of the file without the directory or file suffix.
file_exists: Returns true if a file exists
file_name: Returns the file name from a full path
file_path: Returns the directory/path from a full file path
file_size: Returns a file size
file_suffix: Returns a files suffix/extension
is_directory: Returns true if a file path is a directory
is_file: Returns true if a file path is a file

nyalldawson committed:
Fix broken see link

nyalldawson committed:
[processing] Remove selected rows from batch table, not always last row

epifanio committed:
[processing] Add r.geomorphon grass command and module in grass core plugin (#9625)

* Create r.geomorphon.qgm

* Add files via upload

* Update default.qgc

* Create r.geomorphon.txt

Adding the description file to add the `r.geomorphon` command into `QGIS.Processing.GRASS`

* Update grass7_algorithms_raster_tests.yaml

Adding test for `r.geomorphon`

* fixing wrong spelling

thanks trevis

* Update grass7_algorithms_raster_tests.yaml

adding corrected unit test

* Create r.geomorphon.qgm

* Add files via upload

* Update default.qgc

* Create r.geomorphon.txt

Adding the description file to add the `r.geomorphon` command into `QGIS.Processing.GRASS`

* Update grass7_algorithms_raster_tests.yaml

Adding test for `r.geomorphon`

* fixing wrong spelling

thanks trevis

* Delete CTestCostData.txt

* Delete CTestCostData.txt

* Delete CTestCostData.txt

* Delete CTestCostData.txt

luipir committed:
Update tests/src/python/test_provider_postgres.py

Co-Authored-By: elpaso <elpaso@itopen.it>

nirvn committed:
Add simple line custom dash pattern tests

nirvn committed:
[FEATURE] Make the custom dash pattern dialog a style dock widget

nyalldawson committed:
Fix expression function examples

nirvn committed:
Followup cd5642a: Add dialog title, connect dialog ok/cancel buttons

nirvn committed:
[ui] Use system color for custom dash pattern preview symbol

nirvn committed:
[ui] Improvements to the font marker character widget
- Declare a minimum height to avoid narrow selection area
- Use application color palette instead of hardcoded values
- Decrease number of columns to better fit in the style dock

nirvn committed:
[themes] Add minimum scrollbar dimension to Night Mapping

nirvn committed:
[themes] Tweak palette values to better match designated roles

nirvn committed:
[themes] Set highlight text color palette for themes

nyalldawson committed:
[needs-docs] Tweak custom dash button appearance

- remove "Change" label and replace with larger dash preview icon.
The "change" text is unnecessary and adds to dialog clutter, better
to use the space for a wider preview icon (especially given that
the previous narrow icon never really showed enough of the pattern
to be useful!)

- don't offset the line in the preview if the symbol has an offset
set

- respond correctly to dash pattern, line width unit changes, cap
style changes

- show a nice big preview tooltip on hover

nyalldawson committed:
Fix duplicate layers in custom layer order panel

Fixes #21955

nyalldawson committed:
Avoid ugly super-narrow annotation balloon callout lines at some angles

nyalldawson committed:
Review comments

nyalldawson committed:
[processing] Fix incorrect columns hidden when toggling advanced params in batch mode

nyalldawson committed:
Allow loading QLR files with invalid sources

E.g. if the layer path has moved, we still should allow these
files to be read, so that the layer path can be fixed by the user
manually.

nyalldawson committed:
Disable some layer context menu actions for invalid layers

nirvn committed:
[FEATURE][ui] Add a line edit to type character for the font marker

The line edit also supports pasting of unicode code in decimal or
hexadecimal format (i.e. 123 or 0x2713).

In addition, the character widget now grabs the keyboard
focus and will change the selected character based on
key stroke.

alexbruy committed:
rename method so API is compatible with recent changes and other wrappers

troopa81 committed:
Never use scientific notation

nirvn committed:
Follow up dc7bfc5 : QString::back() too recent to use

nirvn committed:
[FEATURE][symbology] Unlock string as character for font markers

nirvn committed:
[FEATURE][symbology] Unlock string as character for font markers

m-kuhn committed:
Doxygen for QgsSymbolLayer::setRenderingPass

nirvn committed:
[ui] Fix broken text annotation background color

jef-n committed:
query builder: fix #21805 & #21966 & #21967

nyalldawson committed:
Auto emit characterSelected signal

nyalldawson committed:
Allow navigation of character widget with arrow cursor keys

nyalldawson committed:
Don't try to draw characters which don't exist in current font

nyalldawson committed:
Allow QgsScrollArea to maintain horizontal width of area for
child widgets, and apply scroll area to vertical contents only

nyalldawson committed:
Dynamically set number of columns to show in character widget
based on widget width, and avoid unnecessary horizontal scroll bar

nyalldawson committed:
Don't draw background outside of character columns

nyalldawson committed:
Nicer way of ensuring selected character is in view

nyalldawson committed:
Draw missing character backgrounds with different color

nyalldawson committed:
Allow home/end to jump to first/last character

nyalldawson committed:
Allow home/end to jump to first/last character

nyalldawson committed:
Fix resizing of QgsCharacterSelectorDialog

nyalldawson committed:
Nicer behaviour of QgsCharacterSelectorDialog

Commit:
[processing] fix #21958 used sqlite3 module instead of GDAL/ORG's SQLite driver

Commit:
[processing] updated test files - tiles with transparent background

stev-0 committed:
Add boolean type to DB Manager PostGIS provider

vcloarec committed:
Fix crash when open project containing a mesh with a extras dataset groups with incompatible mesh 

QGis application crash when opening a project contain a mesh layer with a dataset groups wich has a incompatible mesh. It can  happen when the dataset groups file change after first loading.

vcloarec committed:
Implementation of QgsMeshLayer::reload() and QgsMdalProvider::reloadData()

Implementation of this override method permit to reload mesh end dataset groups when the data changed outside of the QGIS application.

vcloarec committed:
Add test for QgsMeshLayer::reload() and QgsMdalProvider()::reloadData()

Add test and data file for new test

PeterPetrik committed:
update to MDAL 0.3.2

nirvn committed:
[FEATURE][spatialite] Add a REGEXP function to use for SQL filter et cie

nirvn committed:
Add a REGEXP function in spatialite_connect() python function
This enables DB Manager and the Spatialite Execute SQL algorithm
to use the 'string REGEXP pattern' syntax.

nirvn committed:
Add tests

cjmayo committed:
Install full set of application icons on Linux

Icons were already available in linux/icons directory.
Tidy linux/CMakeLists.txt.

Commit:
[QgsQuick] Fix of value relation widget
Due to last changes in passing style object, styling property has been updated (widget had 0 height without fix).
Fixed case when (current) value is not matching value relation map (it generated error while creating component).

mbernasocchi committed:
fix wrong SVG PATHS logging on server startup

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

rouault committed:
Fix build without WITH_QTWEBKIT

mbernasocchi committed:
remove redundant /profiles when QGIS_CUSTOM_CONFIG_PATH is set

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

rouault committed:
[OGR provider] Fix opening of GeoPackage datasets with foreign key violation

rouault committed:
[GDAL provider] Fix opening of GeoPackage datasets with foreign key violation

nyalldawson committed:
Add method to retrieve map extent to QgsRenderContext

Previously only a "layer clipping extent" was available for retrieval
from a QgsRenderContext instance, yet there's a need for rendering
operations to have access to the original full extent of the map
being rendered.

daniviga committed:
[RPM] Add missing gdal-python-tools dependency
Processing uses some python scripts that are part of gdal-python-tools package

nirvn committed:
[symbology] Fix setSize(size) not working for ellipse markers

nyalldawson committed:
Allow setting order by clause for aggregate calculation

yoichigmf committed:
bug fix   can't open file using Japanese Character in name

nyalldawson committed:
[FEATURE] Add order by support to expression aggregate and concatenation functions

Because certain aggregates and concatenation requires results in
a certain order, this change allows specific control of the order
features are added to the aggregate during an expression evaluation.

E.g.

concatenate("Station",concatenator:=',', order_by:="Station")

will give a comma separated list of station names in alphabetical
order, rather than layer feature order.

Sponsored by SMEC/SJ

nirvn committed:
[ui] Fix marker sizing for the filled marker widget

nyalldawson committed:
[processing] Fix exception when running batch alg after loading parameters

daniviga committed:
[RPM] Add qt5-qtbase-private-devel build dep for F30+
qt5-qtbase-devel as been splitted in Fedora 30

daniviga committed:
[RPM] Fix a typo in a comment in spec file

daniviga committed:
[RPM] Fix a mispelled word

nyalldawson committed:
Fix constraint expression and description are flipped

Fixes #22016

nyalldawson committed:
Split processing test files into multiple batches

There's two motivations for this:
- the existing one was getting massive and took ages to run, which was
a pain when developing. Smaller batches allow just a subset of test to
be run which is much faster.
- There's a random segfault on test exit which occurs on Travis. Rather
then disabling these absolutely critical tests altogether, I'm using
this as a method of bisecting exactly which alg is causing this.

nyalldawson committed:
Spelling, formatting

nyalldawson committed:
Add extra raster hashes

nirvn committed:
[ui] Fix missing disabled state for {check/radio/group}box

nyalldawson commented:
Can you please rebase this on master and I'll investigate the failing test?

spatialthoughts commented:
@nyalldawson Thank you. I have rebased the branch to master. (at least I think I did). Hope you are able to investigate the error.

nirvn committed:
[effects] when creating the preview image, disable effects to avoid 'double apply'

jef-n committed:
fix #22039

nyalldawson committed:
Add missing default value for join style param

nyalldawson committed:
If source crs is not geographic, use WGS84 as a pivot to allow non-geographic sources

nyalldawson commented:
Hmm... I'm stumped with this! I added some fixes but still see different results inside QGIS vs running the test outside QGIS.

luipir commented:
> Hmm... I'm stumped with this! I added some fixes but still see different results inside QGIS vs running the test outside QGIS.

me too :(

luipir commented:
> Hmm... I'm stumped with this! I added some fixes but still see different results inside QGIS vs running the test outside QGIS.

btw now the error is completly changed... expected and result are different (starting from 4th or 5th digit).

roya0045 commented:
Have you tried with another crs? Would it be possible to do a rendering test instead of checking the exact coordinates?

Or a shortcut would be to use the data generated by the test as the expected values, this way we will know if the algorithm's behavior will change in time even if the results are slightly different in Travis.

luipir commented:
> Or a shortcut would be to use the data generated by the test as the expected values, this way we will know if the algorithm's behavior will change in time even if the results are slightly different in Travis.

@roya0045  suppose is the way @spatialthoughts generated test data set... did you @spatialthoughts ?

roya0045 commented:
> > Or a shortcut would be to use the data generated by the test as the expected values, this way we will know if the algorithm's behavior will change in time even if the results are slightly different in Travis.
> 
> @roya0045 suppose is the way @spatialthoughts generated test data set... did you @spatialthoughts ?

Just to be more clear, what I meant by test is to use the travis test results instead of using 'home' results. This is what I do for my rendering tests since I want to make sure that nothing in my setup might generate data that can never be replicated in the branch and build tests, I just use the travis results and check if they are what I expected beforehand.

spatialthoughts commented:
@roya0045 @luipir I generated the expected test data in QGIS running my algorithm. So I suppose there could be a mismatch between what is generated by Travis. How do I use the travis test results as the expected output? I can't figure out how to run this locally and grab the output.

spatialthoughts commented:
I am still looking to merge this. Any suggestions on howto fix the test?

ujavalgandhi committed:
remove tests

ujavalgandhi committed:
remove tests

ujavalgandhi committed:
fix conflicts

ujavalgandhi committed:
Fix type

ujavalgandhi committed:
revert test file

ujavalgandhi committed:
revert test file

luipir commented:
I do not have suggestions other thatn asking to @m-kuhn or @3nids . IMHO cloning this PR without tests in #30150 does not help. There is the rule to avoid merge if not covered by tests.

luipir commented:
I can't reopen this PR because of #30150 of the same branch

DelazJ commented:
And since we are in "feature freeze" period, this new alg can not be merged before release (3.10 is its target)

luipir commented:
> And since we are in "feature freeze" period, this new alg can not be merged before release (3.10 is its target)

imho can be an exception... this is long term PR that have problem to merge (seems) due to testing infrastructure not depending on @spatialthoughts

nyalldawson commented:
I have a concern about allowing this algorithm for non-point layers, because the results will be very misleading. The current approach just creates a CRS around the feature's centroid, so for large lines or polygon objects the results will NOT be accurate.

Given this, I don't feel comfortable allowing anything except single point layers to work with the algorithm -- otherwise we are just making it easy for users to silently create inaccurate results.

luipir commented:
> Given this, I don't feel comfortable allowing anything except single point layers to work with the algorithm -- otherwise we are just making it easy for users to silently create inaccurate results.

@nyalldawson a layer with only a feature or a multi feature layer of a single part point geometry? 

@spatialthoughts can you rebase the code to fix conflicts?

nyalldawson commented:
@luipir 
>a layer with only a feature or a multi feature layer of a single part point geometry?

No, just restricting to any point-based layers only (ie. no line or polygons).

But you raise an interesting point -- we also need to correctly consider multipoints here, by splitting them up, calculating the buffer, and the dissolving all the component buffers back to a single feature.

spatialthoughts commented:
@nyalldawson I tested and geometry.buffer() works on multipoint geometry and does the correct thing by merging the buffers, so this algorithm doesn't need to do this separately. 

Restricting to point layers only sounds good.

I have got this branch in a state that is beyond my skill to rebase. I have a whole bunch of unrelated commits from past attempts and not sure how to proceed. Do you have any guidance? Starting fresh looks easier.

nyalldawson commented:
@spatialthoughts 
> I tested and geometry.buffer() works on multipoint geometry and does the correct thing by merging the buffers, so this algorithm doesn't need to do this separately.

Right - but picture the situation with a multipoint feature, where one point sits on one side of a continent and another on the other side. In this case, it's not appropriate to create a single CRS for use when creating the buffer.

In this situation, you'd need to split the multipoint apart, and create the CRS and buffer for each point separately, and then finally dissolve the individual buffers back to a single multi-polygon object.

nyalldawson commented:
I've tried to fix this branch too, without success. I'd suggest pulling out the files you changed and pasting them into a brand new branch based on master and abandoning this particular PR branch..

spatialthoughts commented:
Closing this PR. Will submit a new PR from a new branch after fixing the issues mentioned earlier.

ghtmtt commented:
@spatialthoughts I think this algorithm is worth to land QGIS. Do you have plans to resurrect it?
